### DISCUSSION TITLE
Could it be that removed columns and data isn't really removed?

### DISCUSSION THREAD

**Discussion by Nasnl** [Budibase/budibase#5404] [General]
I have built an app and have been creating many columns in the process, as well as removing them again. Same with rows.

Now when I do a test and I review the input and output of steps, I see column names and values that should not be in there.

Is there a way to refresh/reset and clean that data that obviously stays present 'under water'?

**Comment by melohagan** [comment]
Hey @Nasnl 

I'm not quite sure what the issue is. Is this for Form Steps?

Would you be able to share some screenshots please?

**Comment by Nasnl** [comment]
Of course @melohagan. Sorry I could not explain it clear enough.

In the attached screenshot of the input for an automation step (coming from the output of the step before), you see fields mentioned like 'From' and 'Sender' (and if you would scroll down, you would also see 'Recipient'). The thing is: I deleted these columns from the table, so they should not be included anymore.
I have been fiddling around quite a lot with testing, so now my automation steps contain many unused columns in the input and output.

I'm trying to pass on an email address through binding to the last step (to send an email), but that doesn't succeed, because the field I have put the email address in is empty, but it does appear in a field I did not create 'PrimaryDisplay' (while I switched off for all columns that they should be 'primary display fields'. I'm not sure if the latter is a separate issue or if it is related to the column names that do not disappear, even though I deleted the columns themselves from the tables.

![ColumnsBudi](https://user-images.githubusercontent.com/8427570/163593927-c7eea405-44c8-468d-8ed1-60c308faeaaf.png)

**Reply by Nasnl** [reply]
To add to this @melohagan: I created a new - completely clean - automation and deleted all rows from all tables. At my first automation test, I entered new data to test with. When I check the output of the first step (trigger at new row created), the data seems correct.

When I check the output of the 2nd step (Query Rows), then the output does not contain the data of my test, but it contains data from the Query Rows output in another automation which has been deleted from the tables already.

This is another example of that I cannot get rid of data I have already deleted; even across automations and deleted tables, obviously.

All of this refers to Budibase Cloud and the internal Budibase database, btw.

**Reply by mjashanks** [reply]
Hi @Nasnl

I think you may have some misunderstanding of how automation work - or it could be me that is misunderstanding your issue :) 

Firstly you said
> ...the input for an automation step (coming from the output of the step before).....The thing is: I deleted these columns from the table

The **input** usually would not contain data from your table. The input is only what the step needs to run. The exception for this is is you are creating or updating a row in a table, therefore the input does include columns in the table.

What type of step are you using in your screenshot.

> When I check the output of the 2nd step (Query Rows), then the output does not contain the data of my test...

The `Query Rows` step fetches data from the database, and **outputs** it. I.e. the output of the `Query Rows` step is the returned rows. Therefore, I would not expect the query rows step to output the data that you input when running "Test Automation" 

That said, if you think that it's returning data that you have already deleted, that would be very odd.

Happy to have a look at your app if you want to export and send to support@budibase.com

**Comment by Nasnl** [comment]
I found another example of this: deleted columns stay present somewhere in the app.

I deleted some columns from my table IncomingMessages to make it simpler and only 4 remained:
subject, IntentLabel, message and Sender

![image](https://user-images.githubusercontent.com/8427570/168388697-d62a2b3e-6b6e-4a55-bedd-5b28ac0ba1d9.png)

When I created a webUI to add a row to that table, it shows me more fields then there actually are in the current table: all columns that I have deleted before are shown in the form:

![image](https://user-images.githubusercontent.com/8427570/168388820-de568d1e-fc5f-4bf0-b54b-5a7eb75794ee.png)

Plus it shows the previous name of the column 'Sender', which used to be 'From'.

I checked several times if it is the same table and if the columns are really deleted. Not mistaken. :-) Also tried in anonymous browser, browser on another machine and after cleaning cache.
